rir/src/compiler/analysis/global_redundancies.h

```diff
@@ -0,0 +1,214 @@
+#ifndef COMPILER_PIR_GR_H
+#define COMPILER_PIR_GR_H
+
+#include "../pir/pir.h"
+#include "generic_static_analysis.h"
+#include <algorithm>
+
+namespace rir {
+namespace pir {
+
+/*
+ * Polynomial Time Global Value Numbering analysis based on:
+ * https://www.sciencedirect.com/science/article/pii/S1477842415300622
+ */
+
+/*
+struct GBNValue {
+    Value* value;
+    SEXP* constant;
+    GBNValue() : value(nullptr), constant(nullptr) {};
+    GBNValue(Value* aValue) : value(aValue), constant(nullptr) {};
+    GBNValue(SEXP* aConstant) : constant(aConstant), value(nullptr) {};
+    
+    bool isValid() const { return value != nullptr || constant != nullptr; } 
+    bool operator==(const GBNValue& anotherGV) {
+        if (value!=nullptr && anotherGV.value!=nullptr) 
+            return value == anotherGV.value;
+        if (constant!=nullptr && anotherGV.constant!=nullptr)
+            return *constant == *anotherGV.constant; 
+        return false;
+    }
+    
+};*/
+
+/*
+ * An abstraction of a set of equivalent expressions.
+ * For more information see page 170. 
+*/
+struct ValueExpression {
+    typedef ValueExpression Expression;
+    ValueExpression() : operand1(), operand2(), operation(nullptr) {};
+    ValueExpression(Value* value) : operand1(value), operand2(), operation(nullptr) {};
+    ValueExpression(Value* value, Value* anotherValue, Tag* operation) : 
+        operand1(value), operand2(anotherValue), operation(operation) {};
+    Value* operand1;
+    Value* operand2;
+    Tag* operation;
+
+    bool contains(Expression& expression) {
+        return (operand1 == expression.operand1) &&
+                (operand2 == expression.operand2) &&
+                (operation == expression.operation); 
+    }
+};
+
+/*
+ * An abstraction of a set of equivalent phi-functions.
+ * For more information see page 170. 
+*/
+struct ValuePhiFunction {
+    ValuePhiFunction(Value* left, Value* right) : leftPath(left), rightPath(right) {};
+    Value* leftPath;
+    Value* rightPath;
+    //BB* joinPoint;
+    bool operator==(const ValuePhiFunction& anotherPF) {
+        return leftPath == anotherPF.leftPath &&
+                rightPath == anotherPF.rightPath;
+    }        
+};
+
+union GRPIRAssignment {
+    SEXP* constant;
+    size_t argIndex;
+    ValueExpression::Expression* expression;
```

Comment:
just `Expression`, no pointer? See above comment on passing pointer to a local variable at `setExpression(&expression)`